subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/LazyPublishArtifact.java

```diff
@@ -80,14 +81,19 @@ public Date getDate() {
 
     private PublishArtifact getDelegate() {
         if (delegate == null) {
-            Object value = provider.get();
+            Object value = provider;
+            while (value instanceof Provider) {
+                value = ((Provider<?>) value).get();
+            }
             if (value instanceof FileSystemLocation) {
                 FileSystemLocation location = (FileSystemLocation) value;
                 delegate = fromFile(location.getAsFile());
             } else if (value instanceof File) {
-                delegate = fromFile((File)value);
+                delegate = fromFile((File) value);
             } else if (value instanceof AbstractArchiveTask) {
-                delegate = new ArchivePublishArtifact((AbstractArchiveTask)value);
+                delegate = new ArchivePublishArtifact((AbstractArchiveTask) value);
+            } else if (value instanceof Task) {
+                delegate = fromFile(((Task) value).getOutputs().getFiles().getSingleFile());
```

Comment:
:x: This is breaking all the dependencies. I.e. the task won't be executed to generate the output. Could you add an integration test for that?

I think there should be a Buildable created here similar to `ArchivePublishArtifact(...)` above that is then used further down to resolve the task dependencies.

Reply:
This code is called with a `TaskProvider`, so the task carries the dependencies. I will add an integration test but as far as I can tell, it just works :tm: , see https://scans.gradle.com/s/sxluuqyhizfq2/timeline

Reply:
Ah ok, because this is already called late when the task dependencies are resolved? Then this is fine.

But then the ArchivePublishArtifact wrapper above looks like it is not needed anymore (that confused me). Can that be remove now as archive tasks can then just be handled as any other task (?).

Reply:
I think it's slightly different, hence why I left the code above: for an archive task, we will specifically look for the `archiveFile` output. It means that if, for some reason, a custom archive task has several outputs, it would still work. This is not the case for the new code which assumes a single output.